### DISCUSSION TITLE
Asking the user if they are sure they want to remove a rule

### DISCUSSION THREAD

**Discussion by nbernardo96** [react-querybuilder/react-querybuilder#402] [Q&A] [answered]
Hello! I was wondering if there is currently any way to prompt the user to confirm if they would like to remove a rule (i.e. via a popup modal or alert). 

I've been looking through the docs/discussions and so far the only solution I came to is to create a custom removeRuleAction and add a function that will display a popup when the user clicks remove. However, the props that are available do not include any details about the rule (I would like to retrieve the value and display it in a popup modal). Thank you!

**Comment by jakeboone02** [comment, marked as answer]
Hi @nbernardo96, thanks for bringing this up. There are a couple of ways we could handle it, one being the way you suggest - adding the `rule` (or `ruleGroup`) to the props for the "remove" buttons. If we did that, we may as well add them to the props for _all_ buttons.

The second way would be to add an `onRemove` callback function prop that returns a `boolean`: `true` for "yes, go ahead and delete this" or `false` for "bail out, don't delete it." It would be similar to the `onAddRule` and `onAddGroup` callbacks.

Actually I'm considering doing both. What do you think?

**Reply by nbernardo96** [reply]
Hi @jakeboone02! Thank you for the quick reply and feedback!

I think adding an `onRemove` callback function prop is a great idea actually and gives more flexibility with handling removing a rule! As for adding the rule props to the remove buttons, I can't currently think of a benefit to having access to the rule prop if the `onRemove` callback function gets implemented (aside from it being nice to have access to should any other use cases pop up that may need it). 

However, if you're considering doing both I think that would be great! :)

**Reply by nbernardo96** [reply]
Actually, I wanted to add to my previous reply that I realized for my use case, adding the rule prop to the remove buttons would be helpful as I plan to use the data associated with the rule (field name, and value). Would `onRemove` have access to the rule prop as well?

**Reply by jakeboone02** [reply]
Yeah, I'm thinking `onRemove` would be passed the rule/group, the `path`, the full `query`, and a `context` that custom "remove" buttons can implement.

I'll have a PR up soon that you can play around with.

**Reply by nbernardo96** [reply]
Awesome! That definitely would work then, looking forward to trying it out!

**Reply by jakeboone02** [reply]
See #403. A CodeSandbox using the build from that PR is here: https://codesandbox.io/s/react-querybuilder-ci-qmqxq1.

**Reply by jakeboone02** [reply]
Oh, forgot to mention--that PR implements both the `onRemove` callback and passes a `ruleOrGroup` prop to all `ActionElement` buttons. You can try both at that CS link.

**Reply by nbernardo96** [reply]
Just checked out the PR as well as the CodeSandbox, they both look great and work really well! Was able to replicate my use case perfectly. 

After testing both methods, I think I prefer/will use the onRemove callback as it seems straightforward for what I need - but both cases worked great!

**Reply by jakeboone02** [reply]
Just published [v5.2.0](https://github.com/react-querybuilder/react-querybuilder/releases/tag/v5.2.0) ([changelog entry](https://github.com/react-querybuilder/react-querybuilder/blob/main/CHANGELOG.md#v520---2022-11-26)) with these additions. If you're happy with it, you can mark this thread as answered.

**Reply by nbernardo96** [reply]
Awesome! Thank you!
